pkg/virt-config/featuregate/feature-gates_test.go

```diff
@@ -0,0 +1,88 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Authors.
+ *
+ */
+
+package featuregate_test
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"kubevirt.io/kubevirt/pkg/virt-config/featuregate"
+)
+
+var _ = Describe("Feature Gate", func() {
+	It("register a simple FG, expect default warning", func() {
+		fg := featuregate.FeatureGate{Name: "my-fg", State: featuregate.GA}
+
+		featuregate.RegisterFeatureGate(fg)
+		DeferCleanup(featuregate.UnregisterFeatureGate, fg.Name)
+
+		Expect(featuregate.FeatureGateInfo(fg.Name)).To(Equal(&featuregate.FeatureGate{
+			Name:    fg.Name,
+			State:   fg.State,
+			Message: fmt.Sprintf(featuregate.WarningPattern, fg.Name, fg.State),
+		}))
+	})
+
+	It("register a FG with an explicit warning", func() {
+		const message = "my-message"
+		fg := featuregate.FeatureGate{Name: "my-fg", State: featuregate.Deprecated, Message: message}
+
+		featuregate.RegisterFeatureGate(fg)
+		DeferCleanup(featuregate.UnregisterFeatureGate, fg.Name)
+
+		Expect(featuregate.FeatureGateInfo(fg.Name)).To(Equal(&featuregate.FeatureGate{
+			Name:    fg.Name,
+			State:   fg.State,
+			Message: message,
+		}))
+	})
+
+	It("register multiple unique FGs", func() {
+		fg1 := featuregate.FeatureGate{Name: "my-fg1", State: featuregate.GA, Message: "my-message"}
+		fg2 := featuregate.FeatureGate{Name: "my-fg2", State: featuregate.GA, Message: "my-message"}
+
+		featuregate.RegisterFeatureGate(fg1)
+		featuregate.RegisterFeatureGate(fg2)
+		DeferCleanup(featuregate.UnregisterFeatureGate, fg1.Name)
+		DeferCleanup(featuregate.UnregisterFeatureGate, fg2.Name)
+
+		Expect(featuregate.FeatureGateInfo(fg1.Name)).To(Equal(&fg1))
+		Expect(featuregate.FeatureGateInfo(fg2.Name)).To(Equal(&fg2))
+	})
+
+	It("register FG that overrides an existing one", func() {
+		fg1 := featuregate.FeatureGate{Name: "my-fg1", State: featuregate.GA, Message: "my-message"}
+		fg2 := featuregate.FeatureGate{Name: "my-fg2", State: featuregate.GA, Message: "my-message"}
+		fg1clone := featuregate.FeatureGate{Name: "my-fg1", State: featuregate.GA, Message: "my-other-message"}
+
+		featuregate.RegisterFeatureGate(fg1)
+		featuregate.RegisterFeatureGate(fg2)
+		featuregate.RegisterFeatureGate(fg1clone)
```

Comment:
Why would we want this?

Reply:
It expresses and documents the begavior when one registers another FG with the same name. This test shows that the behavior is that it overrides the prev registration.

Reply:
Right but isn't this a mistake of the author of the code. A safer way would be to panic in that case.

Reply:
I do not have a strong opinion on this. Panic works for me, but I am just unsure of we should raise a panic in production code.. Having a bug in a FG that was set wrongly is probably better than blowing up.
On the other hand, this should be consistent if such a code exist. I.e. it will panic always and on a very early stage.

I tend to go with a panic, but if someone can add a feedback on this to back this up even more, it will be great.